src/views/Package/PackageState.tsx

```diff
@@ -50,6 +76,165 @@ export const usePackageState = () => {
   return state;
 };
 
+export const appendMenuItem = (
+  items: MenuItem[],
+  item: MenuItem
+): MenuItem[] => {
+  const last = items[items.length - 1];
+
+  if (last && last.level < item.level) {
+    last.children = appendMenuItem(last.children, item);
+    return items;
+  }
+  return [...items, item];
+};
+
+export const splitOnHeaders = (md: string, level: number): string[] => {
+  if (!md) {
+    return [];
+  }
+
+  const regex = new RegExp(`(^#{1,${level}}[^#].*)`, "gm");
+  return (
+    md
+      ?.split(regex)
+      // Trim lines and remove whitespace only entries
+      .reduce((accum: string[], str: string) => {
+        const newStr = str.trim();
+        if (newStr === "") return accum;
+        return [...accum, newStr];
+      }, [])
+  );
+};
+
+const getHeaderAttributes = (hdr: string): { id: string; title: string } => {
+  const attrStrings = hdr.match(/(\S+)\s*=\s*(\"?)([^"]*)(\2|\s|$)/g) ?? [];
+  const attrs: { [key: string]: string } = attrStrings.reduce((accum, str) => {
+    const [key, value] = str.split("=");
+    const [_, parsedValue] = /['"](.*?)['"]/.exec(value) ?? [];
+
+    return {
+      ...accum,
+      [key]: parsedValue,
+    };
+  }, {});
+
+  // Use raw title for items that don't specify data attributes, like readme
+  // headers.
+  const [_, rawTitle] = /^#*\s*([^<]+?)\s*(?:<|$)/.exec(hdr) ?? [];
+  const wEmoji = rawTitle.replace(
+    /:\+1:|:-1:|:[\w-]+:/g,
+    (match: string): string => {
+      return emoji.get(match) ?? match;
+    }
+  );
+  const title: string = attrs["data-heading-title"] ?? wEmoji;
+  const id = attrs["data-heading-id"] ?? encodeURIComponent(sanitize(title));
+
+  return { id, title };
+};
+
+/**
+ * Accept's markdown document from jsii-docgen with readme and api reference
+ * documentation and parses the content into a traversable map of menu items
+ * and types. This allows splitting the rendering of the readme and each item
+ * in the api reference.
+ */
+export const parseMarkdownStructure = (
+  input: string,
+  {
+    scope,
+    language,
+    name,
+    version,
+  }: { scope?: string; language: string; name: string; version: string }
+): { readme: string; apiReference: Types; menuItems: MenuItem[] } => {
+  const nameSegment = scope ? `${scope}/${name}` : `${name}`;
+  const basePath = `/packages/${nameSegment}/v/${version}`;
+  const langQuery = `?${QUERY_PARAMS.LANGUAGE}=${language}`;
+  const separator =
+    '\n# API Reference <span data-heading-title="API Reference" data-heading-id="api-reference"></span>\n';
```

Comment:
If the package's README is empty or non-existent, will there still be a newline at the beginning? If not, this separator might be found zero times (even though there is a "API Reference" heading in the file), and I think there's a possibility we could end up with a bug where "API Reference" appears twice in the table of contents.

Reply:
removed the leading `\n`. I believe we can rely on it's presence as much as we can rely on any of the structure within the document returned by jsii-docgen (which in this functionality is obviously a lot), as we join all segments of the larger md document with `\n` and the readme is always it's own segment. So at minimum we should have `readmetext\napireference` if not `readmetext\n\napireference`. Regardless we don't need it and its one less assumption so 👍🏻